src/psyclone/tests/psyad/tl2ad_test.py

```diff
@@ -0,0 +1,108 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''A module to perform pytest tests on the code in the tl2ad.py file
+within the psyad directory.
+
+'''
+from __future__ import print_function, absolute_import
+import logging
+import pytest
+
+from psyclone.psyir.frontend.fortran import FortranReader
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.psyad import generate
+
+
+# generate function
+def test_generate():
+    '''Test that the generate() function works as expected.'''
+
+    tl_fortran_str = (
+        "program test\n"
+        "integer :: a\n"
+        "a = 0.0\n"
+        "end program test\n")
+    expected_ad_fortran_str = (
+        "program test\n"
+        "  integer :: a\n\n"
+        "  a = 0.0\n\n"
+        "end program test\n")
+    reader = FortranReader()
+    tl_psyir = reader.psyir_from_source(tl_fortran_str)
+
+    writer = FortranWriter()
+    ad_fortran_str = writer(ad_psyir)
+    assert expected_ad_fortran_str in ad_fortran_str
+
+
+# generate function logging
+@pytest.mark.xfail(reason="issue #1235: caplog returns an empty string in "
+                   "github actions and some flavours of Python", strict=False)
+def test_generate(caplog):
+    '''Test that logging works as expected in the generate() function.'''
+
+    tl_fortran_str = (
+        "program test\n"
+        "integer :: a\n"
+        "a = 0.0\n"
+        "end program test\n")
+    expected_ad_fortran_str = (
+        "program test\n"
+        "  integer :: a\n\n"
+        "  a = 0.0\n\n"
+        "end program test\n")
+    reader = FortranReader()
+    tl_psyir = reader.psyir_from_source(tl_fortran_str)
+
+    with caplog.at_level(logging.INFO):
+        ad_psyir = generate(tl_psyir)
+        assert caplog.text == ""
+
```

Comment:
Should have a check here that the expected logging output has been created?

Reply:
I'm not sure I understand the comment I'm afraid. This does test the logging output with the `assert caplog.text == ""` line. It just happens that there is no logging output when the logging level is set to INFO.

Reply:
No, I don't understand the comment either!